beacon-chain/p2p/utils.go

```diff
@@ -123,43 +124,130 @@ func privKeyFromFile(path string) (*ecdsa.PrivateKey, error) {
 
 // Retrieves node p2p metadata from a set of configuration values
 // from the p2p service.
-// TODO: Figure out how to do a v1/v2 check.
+// When using static peer id, metaDataFromConfig returns default V1 metadata.
 func metaDataFromConfig(cfg *Config) (metadata.Metadata, error) {
-	defaultKeyPath := path.Join(cfg.DataDir, metaDataPath)
-	metaDataPath := cfg.MetaDataDir
+	// Load V1 metadata by default, since V1 metadata can be covered to V0.
+	defaultMd := &pb.MetaDataV1{
+		SeqNumber: 0,
+		Attnets:   bitfield.NewBitvector64(),
+		Syncnets:  bitfield.NewBitvector4(),
+	}
+	wrappedDefaultMd := wrapper.WrappedMetadataV1(defaultMd)
 
-	_, err := os.Stat(defaultKeyPath)
-	defaultMetadataExist := !os.IsNotExist(err)
-	if err != nil && defaultMetadataExist {
-		return nil, err
+	// If --p2p-static-id is false, return default metadata for initialization
+	if !cfg.StaticPeerID {
+		return wrappedDefaultMd, nil
 	}
-	if metaDataPath == "" && !defaultMetadataExist {
-		metaData := &pb.MetaDataV0{
-			SeqNumber: 0,
-			Attnets:   bitfield.NewBitvector64(),
-		}
-		dst, err := proto.Marshal(metaData)
+
+	mdPath, exist := resolveMetaDataPath(cfg)
+	if exist {
+		md, err := metaDataFromFile(mdPath)
 		if err != nil {
+			if errors.Is(err, ssz.ErrSize) {
+				// In case previous metadata file is encoded by proto,
+				// we need to migrate it into ssz encoded version.
+				return migrateFromProtoToSsz(mdPath)
+			}
 			return nil, err
 		}
-		if err := file.WriteFile(defaultKeyPath, dst); err != nil {
-			return nil, err
+		return md, err
+	}
+	if err := saveMetaDataToFile(mdPath, wrappedDefaultMd); err != nil {
+		return nil, err
+	}
+
+	return wrappedDefaultMd, nil
+}
+
+// resolveMetaDataPath returns path and the existence of that path.
+// Issue while opening a file(e.g. permission issues) will be handled at metaDataFromFile.
+func resolveMetaDataPath(cfg *Config) (string, bool) {
+	var mdPath string
+
+	// Prioritize if --p2p-metadata is provided.
+	if cfg.MetaDataDir != "" {
+		mdPath = cfg.MetaDataDir
+	} else {
+		mdPath = path.Join(cfg.DataDir, metaDataPath)
+	}
+
+	// Return path and existence of the file.
+	_, err := os.Stat(mdPath)
```

Comment:
initial review feels like we should at least have some kind of log if it errors for something that's not exist